mongo/legacy.go

```diff
@@ -0,0 +1,301 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package mongo
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/network"
+	"github.com/juju/juju/service"
+	"github.com/juju/juju/service/common"
+)
+
+// TODO(juju4): remove support for upgrading from older mongos
+
+// maybeUseLegacyMongo returns nil if there's a juju that's running
+// on a mongo that's not from the juju-db snap.
+// This is used to preserve the setup up older versions being upgraded.
+func maybeUseLegacyMongo(args EnsureServerParams, search SearchTools) error {
+	if args.DataDir == "" {
+		args.DataDir = "/var/lib/juju"
+	}
+	mongoPath, mongodVersion, err := findLegacyMongo(search)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// As a safety check, if there's no mongo installed but there's
+			// database files present, that's an issue since we can only use
+			// an older mongo to read them..
+			if search.Exists(dbDir(args.DataDir)) {
+				return errors.New("mongo database files exist but no mongo is installed")
+			}
+		}
+		return errors.Trace(err)
+	}
+	// Coming from an early trusty install is not supported since
+	// we now only support wired tiger.
+	if mongodVersion.Major == 2 {
+		return errors.NotSupportedf("mongo %v", mongodVersion)
+	}
+	logVersion(mongoPath)
```

Comment:
The logic from here on is essentially copied from what was there in juju 2.9 to support mongo from apt as a systemd service.